md4c/md4c.c

```diff
@@ -3398,6 +3415,47 @@ md_resolve_links(MD_CTX* ctx, const MD_LINE* lines, int n_lines)
             continue;
         }
 
+        /* Detect and resolve wiki links. */
+        if ((ctx->parser.flags & MD_FLAG_WIKILINKS) &&
+            next_opener != NULL && next_closer != NULL &&
+            (opener->end - opener->beg == 1) &&
+            (next_opener->beg == opener->beg - 1) &&
+            (next_closer->beg == closer->beg + 1) &&
+            (next_opener->end - next_opener->beg == 1) &&
+            (next_closer->end - next_closer->beg == 1) &&
+            (next_opener->ch == '[' && next_closer->ch == ']'))
+        {
+
+            is_link = md_is_inline_wikilink_spec(ctx, lines, opener->end);
+            if (!is_link) {
+                opener_index = next_index;
+                continue;
+            }
+
+            opener->beg = next_opener->beg;
+            closer->end = next_closer->end;
+
+            /* This does not seem to do much: */
+            /* next_opener->ch = 'D'; */
```

Comment:
Do you have some comment on this? There is something that I'm not understanding. I was expecting the wiki link detection not to work without it, but it does work.

Reply:
I believed it is needed for some obscure inputs like `[[foo]](bar)`, thinking it goes roughly as follows:

1. `md_analyze_links()` is called for the inner `[...]` first. 
2. You check whether it is enclosed in another `[...]`. It is -> hence you resolve as a wikilink.
3. Then `md_analyze_links()` is called for the outer `[...]`.
4. It sees it makes an inline links (with the `bar` as its destination). The link demolishes everything detected inside of it and then analyzes its text as sort of a standalone paragraph from scratch, killing the previous wiki-link detection.

Clearly, there is something else what makes it all play differently. Maybe I overlooked/forgot something... Needs more analysis.

Reply:
I think the link nesting rules save it here. CommonMark specifies the links cannot be nested in each other (and if they are, only the innermost is really a link). This is handled in `md_analyze_links()` with this piece of code:

``` C
        /* If nested ("[ [ ] ]"), we need to make sure that:
         *   - The outer does not end inside of (...) belonging to the inner.
         *   - The outer cannot be link if the inner is link (i.e. not image).
         *
         * (Note we here analyze from inner to outer as the marks are ordered
         * by closer->beg.)
         */
        if((opener->beg < last_link_beg  &&  closer->end < last_link_end)  ||
           (opener->beg < last_img_beg  &&  closer->end < last_img_end)  ||
           (opener->beg < last_link_end  &&  opener->ch == '['))
        {
            opener_index = next_index;
            continue;
        }
```

Therefore the outer ones are not really considered.

Reply:
But it is a different story with an image:
```
![[foo]](url)
```
translates into
```
<p><img src="url" alt="[foo]"></p>
```

Here the reason is that when handling the inner pair in `md_analyze_links()`, the `next_opener` points `MD_MARK` describing the image opener `![` and it has `next_opener->ch == '!'` therefore the big `if` for the wiki-link detection fails.

The question is what precedence rules we ant to here.

Anyway, I would remove those two lines, even if we decode to change the precedence to prefer wiki-link here, it would have to be done differently.